keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1304 @@
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare zero trust policies is
+important to compliance. i.e., it should be possible for policies governing 
+resources like namespace and roles to default to the most restrictive state
+when first created.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to webhook Rules, NamespaceSelector,
+  ObjectSelector)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  and
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to webhooks MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+severity, levels, failure policies, type safety, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  kind:
+    group: apps
+    kind: Deployment
+    version: v1
+  validations:
+    - expression: "self.spec.replicas <= config.maxReplicas"
+      # ...other rule related fields here...
+```
+
+The `spec.config` field of the `ValidatingAdmissionPolicy` references the
+"configuration CRD" for this admission policy. For this example, the
+configuration CRD is `ReplicaLimit`.
+
+Note: This is a "Bring Your Own CRD" design. The admission policy definition
+author is responsible for providing the `ReplicaLimit` configuration CRD.
```

Comment:
+1 to figuring out how best to deal with the shared ownership here -- to the extent that the schema for the configuration CRD is configuring functionality implemented by K8s, figuring out how to deal with issues like version skew will be important.

Reply:
After looking more at version skew issues, CRD issues and aggregated API server issues, I'm incline to make CEL dynamic, we can still run the type checker but it will only be informational (generates warnings). This helps with part of the problem. The other thing we can do is use matchPolicy, I've added "Use Case: Multiple policy definitions for different versions of CRD" to show an example of this idea.